### PR TITLE
Add Input Validation for CLV Models

### LINKED ISSUE #175: Add `_check_inputs()` Utility to `clv` Module?
`lifetimes` has a utility function for validating inputs prior to model fitting, which I expanded on in my `btyd` fork for `GammaGammaModel` input validation:

https://github.com/ColtAllen/btyd/blob/main/btyd/models/__init__.py#L254

(note that L294-L297 should be removed because I've since realized this is not a modeling assumption. In fact, rescaling time can help with model convergence when fitting to many years of data).

**`logp` in `BetaGeoModel` and the `clv.distribution` classes are already checking for many of these constraints, but I'm not sure how informative the error messages are. Should we add an additional validation check utility?**

Usually these constraints won't be violated if data is preprocessed with `clv_summary()` prior to model fitting, but that function is in written in `pandas` and will not be adequate for massive datasets exceeding local RAM size, particularly for data scientists using PySpark:

https://www.databricks.com/notebooks/Customer%20Lifetime%20Value%20Virtual%20Workshop/02%20The%20BTYD%20Models.html

Or just querying directly from a database, which is why the documentation for my `btyd` fork provides a SQL template:

https://btyd.readthedocs.io/en/latest/Generate%20RFM%20Summaries%20From%20a%20SQL%20Store.html

Could be worth referencing both of those in the documentation.

### PR DESCRIPTION
**Pull request by shivamlalakiya**
## Description
Addresses issue #175.

Following maintainer feedback, this PR implements a shared `_check_inputs` utility method in the base `CLVModel` rather than limiting it to `BetaGeoModel`. This centralizes data validation (`frequency`, `recency`, `T`) for all models that share identical input requirements (BG/NBD, Pareto/NBD, MBG/NBD). 

It also includes parameter handling to support the specific condition variations required by models like `ShiftedBetaGeoModel`.

Previously, invalid data (e.g., negative inputs or `recency > T`) would cause obscure errors downstream during model fitting or sampling. Now, the base model catches these early and raises a clear, descriptive `ValueError`.

## Changes
* Added `_check_inputs` to the base `CLVModel` to validate `frequency`, `recency`, and `T`.
* Added parameter support to accommodate specific edge cases (e.g., for `ShiftedBetaGeoModel`).
* Updated relevant CLV models to utilize this shared base validation.
* Updated unit tests in `tests/clv/` covering:
    * Negative values for frequency, recency, and T.
    * The invalid case where `recency > T`.
    * Ensure base model validation logic covers multiple CLV models.

## Related Issue
- [x] Closes #175 
- [ ] Related to #

## Checklist
- [x] Checked that [the pre-commit linting/style checks pass](https://www.pymc-marketing.io/en/latest/contributing/index.html). Feel free to comment [`pre-commit.ci autofix` to auto-fix](https://pre-commit.ci/#configuration-autofix_prs).
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks) using [numpydoc format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [x] If you are a pro: each commit corresponds to a [relevant logical change]

----
📚 Documentation preview 📚: https://pymc-marketing--2305.org.readthedocs.build/en/2305/

### CODE REVIEW

File: test_model_multi
Comment (ColtAllen) [inline test_model_multi:1]: Let's remove this.

**Review by ColtAllen — CHANGES_REQUESTED** [review]
Hey @shivamlalakiya,

Thanks for opening this PR! Can we move this method into `CLVModel` so that all CLV models can use it? Supporting the data requirements of all models will require some modification, but the end result will look very similar to this [legacy code](https://github.com/ColtAllen/btyd/blob/main/btyd/models/__init__.py#L254).

I'm nearly done with a major PR involving changes to the CLV API. For this PR we can keep things simple and just make changes to `CLVModel`, and I'll update my PR with the model-specific validation calls.

**Review by ColtAllen — CHANGES_REQUESTED** [review]
Keep this in `CLVModel` despite what the Bugbot says. Yes, some changes are needed  to support `ShiftedBetaGeoModel`, but we can add a new parameter for those conditions. The BG/NBD, Pareto/NBD, and MBG/NBD all have identical requirements.

**Review by ColtAllen — CHANGES_REQUESTED** [review]
Thanks @shivamlalakiya - this PR is on the right track! Can we add a monetary_value parameter for the Gamma-Gamma model as well? Tests will also need to be added for models other than BGNBD and sBG.

File: test_model_multi
Comment (ColtAllen) [reply test_model_multi:1]: Curious how this file keeps getting re-created? Is this due to an MMM test?

**Comment by shivamlalakiya** [comment]
Thank you @ColtAllen , I have updated the PR to address all the feedback. Let me know if there's anything else you'd like me to tweak before we merge!

File: pymc_marketing/clv/models/modified_beta_geo.py [hunk 1]
```diff
@@ -283,6 +284,29 @@ def build_model(self) -> None:  # type: ignore[override]
...
+            data["monetary_value"] = monetary_value
+
+        return customer_lifetime_value(
+            transaction_model=self,
+            data=data,
+            future_t=future_t,
+            discount_rate=discount_rate,
+            time_unit=time_unit,
+        )
+
```
Comment (PabloRoque) [inline pymc_marketing/clv/models/modified_beta_geo.py]: Why do we need this here? Are we overwriting anything on `BetaGeoModel`? Or is it the same method?

Comment (shivamlalakiya) [reply pymc_marketing/clv/models/modified_beta_geo.py]: @PabloRoque Thank you, ModifiedBetaGeoModel inherits directly from CLVModel, not from BetaGeoModel, so nothing is being overridden here. This method is needed — without it, ModifiedBetaGeoModel would have no way to call expected_customer_lifetime_value directly. That said, since the implementation is identical to ParetoNBDModel's, would you prefer I move it up to CLVModel as a shared base method?

Comment (ColtAllen) [reply pymc_marketing/clv/models/modified_beta_geo.py]: > ModifiedBetaGeoModel would have no way to call expected_customer_lifetime_value directly

None of the transaction models call `expected_customer_lifetime_value` directly. It is only called from the `GammaGammaModel` for monetary value, which itself is a wrapper for `clv.utils.customer_lifetime_value`, which users can call directly if they want to use monetary predictions from an external model.

Comment (PabloRoque) [reply pymc_marketing/clv/models/modified_beta_geo.py]: @shivamlalakiya I am having a look at this line `class ModifiedBetaGeoModel(BetaGeoModel)` so indeed inherits from `BetaGeoModel` and that method is described in that class if I recall correctly.

**Review by ColtAllen — CHANGES_REQUESTED** [review]
Remove the CLV method from the MBG/NBD model.

I noticed the docstrings for the CLV estimation functions are outdated, which might be the source of the confusion. While you're at it, do you mind quickly updating these two lines to indicate the MBG/NBD model is supported as well?

https://github.com/pymc-labs/pymc-marketing/blob/main/pymc_marketing/clv/models/gamma_gamma.py#L175

https://github.com/pymc-labs/pymc-marketing/blob/main/pymc_marketing/clv/utils.py#L65

File: pymc_marketing/clv/models/basic.py [hunk 2]
```diff
@@ -90,6 +91,66 @@ def _validate_cols(
...
+
+        if np.isnan(recency).any() or np.isnan(T).any():
+            raise ValueError("Missing values (NaN/pd.NA) found in recency or T.")
+
+        if frequency is not None:
+            frequency = np.asarray(frequency).astype(float)
+            if np.isnan(frequency).any():
+                raise ValueError("Missing values (NaN/pd.NA) found in frequency.")
+            if np.any(frequency < 0):
+                raise ValueError("Frequency must be >= 0.")
```
Comment (PabloRoque) [inline pymc_marketing/clv/models/basic.py]: This is not the case for the `ModifiedBetaGeoModel`. That model can produce outputs without repeated purchases.

File: pymc_marketing/clv/models/beta_geo.py [hunk 3]
```diff
@@ -178,6 +178,11 @@ def __init__(
             ],
             must_be_unique=["customer_id"],
         )
+        self._check_inputs(
+            recency=data["recency"],
+            T=data["T"],
+            frequency=data["frequency"],
+        )
```
Comment (PabloRoque) [inline pymc_marketing/clv/models/beta_geo.py]: Maybe we can move `_check_inputs` to `CLVModel.__init__`? We are copying boiler plate in a number of models, and I think this could be handled there at once.

File: pymc_marketing/clv/models/beta_geo_beta_binom.py [hunk 4]
```diff
@@ -170,6 +170,11 @@ def __init__(
             must_be_unique=["customer_id"],
             must_be_homogenous=["T"],
         )
+        self._check_inputs(
+            recency=data["recency"],
+            T=data["T"],
+            frequency=data["frequency"],
+        )
```
Comment (PabloRoque) [inline pymc_marketing/clv/models/beta_geo_beta_binom.py]: `CLVModel.__init__`?

File: pymc_marketing/clv/models/gamma_gamma.py [hunk 5]
```diff
@@ -190,18 +191,27 @@ def expected_customer_lifetime_value(
...
+
+        if monetary_value is None:
+            # Use Gamma-Gamma estimates for the expected_spend values
+            predicted_monetary_value = self.expected_customer_spend(data=data)
+            data["future_spend"] = predicted_monetary_value.mean(
+                ("chain", "draw")
+            ).copy()
+        else:
+            # Use the explicitly provided monetary_value
+            data["future_spend"] = monetary_value
```
Comment (PabloRoque) [inline pymc_marketing/clv/models/gamma_gamma.py]: I don't undestand the logic here. Do you want to overide what the model provides? Then I think you can do it externally to the library.

File: pymc_marketing/clv/models/modified_beta_geo.py (hunk 1 above)
Comment (PabloRoque) [reply pymc_marketing/clv/models/modified_beta_geo.py]: @shivamlalakiya I understand now. `expected_customer_lifetime_value` is an external helper, not attached to any of the classes. Since that method accepts `transaction_model` then I don't see the need to add this.

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 6]
```diff
@@ -225,6 +225,11 @@ def __init__(
             ],
             must_be_unique=["customer_id"],
         )
+        self._check_inputs(
+            recency=data["recency"],
+            T=data["T"],
+            frequency=data["frequency"],
+        )
```
Comment (PabloRoque) [inline pymc_marketing/clv/models/pareto_nbd.py]: `CLVModel.__init__`?

**Review by PabloRoque — COMMENTED** [review]
@shivamlalakiya Thanks for the contribution, and sorry for the confusion around `expected_customer_lifetime_value`. I left some comments on the PR. Let me know your thoughts.

**Review by ColtAllen — COMMENTED** [review]
@PabloRoque what do you think the code logic would look like if this were in `__init__.py`? I agree the current approach of parametrizing specific requirements in each model is a code smell.

Another possibility could be to use structured pattern matching with the `_model_type` attribute of the model child classes. This should require only a single parameter for `_check_inputs`, and would look something like this pseudocode:

```python

# declaration in CLVModel
def check_inputs(self, model_type):
     match model_type:
         case ["BG/NBD", "MBG/NBD", "Pareto/NBD":
             # self.data["column"] checks
         case ""Gamma-Gamma Model (Individual Transactions)"":
             # frequency and monetary checks
         case "BG/BB":
             # input checks
         case "sBG":
             # input checks
         case _:
            raise ValueError("Invalid Model Type") # this will probably never happen, but we still need a wildcard case

# usage in child class
self._check_inputs(self._model_type).
``` 

This way validation logic for all models is hard-coded and lives in this single method.

File: pymc_marketing/clv/models/basic.py (hunk 2 above)
Comment (ColtAllen) [reply pymc_marketing/clv/models/basic.py]: This looks like a non-negativity check, but yes - `frequency >= 1` is only required for `GammaGammaModel`.

File: tests/clv/models/test_pareto_nbd.py [hunk 7]
```diff
@@ -89,6 +88,74 @@ def setup_class(cls):
...
+            ParetoNBDModel(data=invalid_data)
+
+    def test_expected_customer_lifetime_value_removed(self):
+        """ParetoNBDModel no longer exposes expected_customer_lifetime_value."""
+        dummy_data = pd.DataFrame(
+            {"customer_id": [1], "frequency": [2], "recency": [5], "T": [10]}
+        )
+        model = ParetoNBDModel(data=dummy_data)
+        assert not hasattr(model, "expected_customer_lifetime_value")
+
```
Comment (ColtAllen) [inline tests/clv/models/test_pareto_nbd.py]: This test should be removed.

**Review by ColtAllen — CHANGES_REQUESTED** [review]
@shivamlalakiya we need to address the points raised by @PabloRoque and myself in https://github.com/pymc-labs/pymc-marketing/pull/2305#pullrequestreview-3859128343. The "Ask" mode in whatever AI you're coding with will give some pointers.

**Before writing any more code** We have a [high-priority issue](https://github.com/pymc-labs/pymc-marketing/issues/2270) and a CLV API change PR about to be posted, both of which may create merge conflicts with your PR. Let's get both of those merged first, but in the meantime let us all strategize the most Pythonic path forward.

**Comment by ColtAllen** [comment]
Hey @shivamlalakiya, sorry this fell off my radar. https://github.com/pymc-labs/pymc-marketing/issues/2270 is merged now and work on this can continue. Let me know if you need help resolving the merge conflicts.

> Location: Should this match/case block live directly inside CLVModel._check_inputs(), or would you both prefer it somewhere else in the module (like init.py)?

Let's keep it in `CLVModel._check_inputs()`, though maybe give it a more descriptive name like `_validate_rfm_data`.

> Scope: Should I apply this match/case refactor to all models right now (including Gamma-Gamma and ShiftedBetaGeoModelIndividual), or just start with the primary transaction models (BG/NBD, MBG/NBD, Pareto/NBD, sBG, BG/BB)?

Let's apply it to all models because the additional effort required is minor. Let me know if you need help with the data requirements of each model.

> Syntax: I plan to match on the exact _model_type strings (e.g., "Shifted Beta-Geometric"). Does that work for you?

Yes.

**Comment by shivamlalakiya** [comment]
Includes an unrelated CI fix for test_log_model_graph_no_graphviz (MLflow 3.x log ordering flake) to unblock test groups 4/5. Let me know if we need to revert this.

**Comment by shivamlalakiya** [comment]
CI note: the only failing job is `tests/test_mlflow.py::test_autolog_pymc_model[blackjax]` (`TypeError: build_kernel.<locals>.kernel() got an unexpected keyword argument 'progress_bar'`). This is unrelated to this PR — it's blackjax dependency drift that also fails on other open PRs (e.g. #2742) and would fail on `main` as well (`main` was last green on 2026-07-05, the exact commit this branch is rebased onto). This PR touches no mlflow/sampling code, and all CLV tests — including the slow-sampling jobs — pass. Should clear once blackjax is pinned/fixed on `main`.

**Comment by ColtAllen** [comment]
Thanks for resurrecting this @shivamlalakiya ! 

> CI note: the only failing job is `tests/test_mlflow.py::test_autolog_pymc_model[blackjax]` (`TypeError: build_kernel.<locals>.kernel() got an unexpected keyword argument 'progress_bar'`). 

This was resolved in the `v1.0.0` branch, which is the branch for all PRs included in the 1.0 release. Can you switch from `main` to that branch and see if there are any merge conflicts?

**Comment by shivamlalakiya** [comment]
@ColtAllen thanks — rebasing onto `v1.0.0` did the trick. I've retargeted the PR base to `v1.0.0`, and **all 54 checks are green**. The blackjax NUTS failure is gone now that we're on `v1.0.0` (`blackjax<1.6`).

The only merge conflict was the removed `ShiftedBetaGeoModelIndividual` (#2734) plus two shifted-model tests that needed updating to the new `build_model(data=...)` API — both resolved.

Quick recap of how this now addresses the earlier review from you and @PabloRoque:

- **Centralized in `CLVModel`**: `_validate_rfm_data` / `_validate_frequency` are shared by BG/NBD, Pareto/NBD, and MBG/NBD; Gamma-Gamma adds its monetary-value + frequency checks; ShiftedBetaGeo keeps its `recency`/`T` constraints. The shared frequency check is non-negativity only, so MBG's zero-frequency customers are still valid.
- Filled the per-model `_validate_data` placeholders `v1.0.0` left behind (the `# ... superceded by #2305` ones) via `super()._validate_data()`; validation runs inside `build_model` (that's where the data lives now).
- Removed the redundant `ModifiedBetaGeoModel` override, removed the stray `test_model_multi`, dropped the old per-model validation tests, and updated the "supported models" docstrings in `gamma_gamma.py` and `clv/utils.py` to include MBG/NBD.
- Consolidated the tests into `tests/clv/test_models.py`, covering every model with `match=` on the specific error messages.

One thing I'd like your call on: I implemented the dispatch as per-model `_validate_data` overrides (matching the placeholders already in `v1.0.0`) rather than the earlier `match/case` on `_model_type`. Happy to switch to `match/case` if you'd prefer.

Does this look good to you, or is there anything else you'd like changed before merge?

**Comment by shivamlalakiya** [comment]
Hi @ColtAllen — this PR was closed automatically on 2026-07-28, not by a review decision. The `v1.0.0` branch was deleted right after #2677 was squash-merged into `main`, and GitHub auto-closes every PR whose base branch is removed. Seven PRs went out the same way, including #2735, #2724, #2630, #2754 and #2742.

Some context on how it ended up targeting `v1.0.0`: it was originally against `main`, and I retargeted it on 2026-07-18 per your suggestion to get around the `blackjax` NUTS failure. After the rebase all 54 checks were green and the last outstanding item was the `match/case` placement, which we settled on keeping in `CLVModel`.

The branch is intact at `209f84be`. Since `v1.0.0` was squash-merged rather than merged, I'd rebase with `git rebase --onto upstream/main <old-v1.0.0-head>` so only my commits replay onto `main`.

Could you reopen and retarget the base to `main`? GitHub disables the reopen button while the base ref is missing, so it needs to be switched first. If you'd rather have a clean PR against `main` instead, happy to open one — just let me know which you prefer.

Same message works for 2304 with @ricardoV94/@williambdean, branch 3e5b363e, and "the constant-tensor DiracDelta approach ricardoV94 signed off on" swapped in for the match/case line. Say the word and I'll write that one too.

**Comment by ColtAllen** [comment]
Sorry about that @shivamlalakiya , deleting the remote `v1.0.0` branch impacted a PR I was working on as well.

Can you reopen a clean PR from `main`? I'd like for you to have credit for this work.
